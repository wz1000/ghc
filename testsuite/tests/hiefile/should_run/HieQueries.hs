{-# LANGUAGE ScopedTypeVariables #-}
module Main where

import System.Environment

import GHC.Types.Name.Cache
import GHC.Types.SrcLoc
import GHC.Types.Unique.Supply
import GHC.Types.Name
import Data.Tree
import GHC.Iface.Ext.Binary
import GHC.Iface.Ext.Types
import GHC.Iface.Ext.Utils
import Data.Maybe (fromJust)
import GHC.Driver.Session
import GHC.SysTools
import GHC.Utils.Outputable                 ( Outputable, renderWithStyle, ppr, defaultUserStyle, initSDocContext, text)
import qualified Data.Map as M
import Data.Foldable

class C a where
  f :: a -> Char

instance C Char where
  f x = x

instance C a => C [a] where
  f x = 'a'

foo :: C a => a -> Char
foo x = f [x]
--      ^ this is the point
point :: (Int,Int)
point = (31,9)

bar :: Show x => x -> String
bar x = show [(1,x,A)]
--      ^ this is the point'
point' :: (Int,Int)
point' = (37,9)

data A = A deriving Show

makeNc :: IO NameCache
makeNc = do
  uniq_supply <- mkSplitUniqSupply 'z'
  return $ initNameCache uniq_supply []

dynFlagsForPrinting :: String -> IO DynFlags
dynFlagsForPrinting libdir = do
  systemSettings <- initSysTools libdir
  return $ defaultDynFlags systemSettings (LlvmConfig [] [])

main = do
  libdir:_ <- getArgs
  df <- dynFlagsForPrinting libdir
  nc <- makeNc
  hfr <- readHieFile (NCU (\f -> pure $ snd $ f nc)) "HieQueries.hie"
  let hf = hie_file_result hfr
      refmap = generateReferencesMap $ getAsts $ hie_asts hf
  explainEv df hf refmap point
  explainEv df hf refmap point'
  return ()

explainEv :: DynFlags -> HieFile -> RefMap Int -> (Int,Int) -> IO ()
explainEv df hf refmap point = do
  putStrLn $ replicate 26 '='
  putStrLn $ "At point " ++ show point ++ ", we found:"
  putStrLn $ replicate 26 '='
  putStr $ drawForest ptrees
  where
    trees = getEvidenceTreesAtPoint hf refmap point

    ptrees = fmap (pprint . fmap expandType) <$> trees

    expandType = text . renderHieType df .
      flip recoverFullType (hie_types hf)

    pretty = unlines . (++["└"]) . ("┌":) . map ("│ "++) . lines

    pprint = pretty . renderWithStyle (initSDocContext df sty) . ppr
    sty = defaultUserStyle
