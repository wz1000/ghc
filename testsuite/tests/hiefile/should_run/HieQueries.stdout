==========================
At point (31,9), we found:
==========================
┌
│ $dC at HieQueries.hs:31:1-13, of type: C [a]
│     is an evidence variable bound by a let, depending on: [$fC[], $dC]
│           with scope: LocalScope HieQueries.hs:31:1-13
│           bound at: HieQueries.hs:31:1-13
│     Defined at <no location info>
└
|
+- ┌
|  │ $fC[] at HieQueries.hs:(4,1)-(82,26), of type: forall a. C a => C [a]
|  │     is an evidence variable bound by an instance
|  │           with scope: ModuleScope
|  │           
|  │     Defined at HieQueries.hs:27:10
|  └
|
`- ┌
   │ $dC at HieQueries.hs:31:1-13, of type: C a
   │     is an evidence variable bound by a type signature
   │           with scope: LocalScope HieQueries.hs:31:1-13
   │           bound at: HieQueries.hs:31:1-13
   │     Defined at <no location info>
   └

==========================
At point (37,9), we found:
==========================
┌
│ $dShow at HieQueries.hs:37:1-22, of type: Show [(Integer, x, A)]
│     is an evidence variable bound by a let, depending on: [$fShow[],
│                                                            $dShow]
│           with scope: LocalScope HieQueries.hs:37:1-22
│           bound at: HieQueries.hs:37:1-22
│     Defined at <no location info>
└
|
+- ┌
|  │ $fShow[] at HieQueries.hs:(4,1)-(82,26), of type: forall a. Show a => Show [a]
|  │     is an evidence variable bound by an instance
|  │           with scope: ModuleScope
|  │           
|  │     Defined in `GHC.Show'
|  └
|
`- ┌
   │ $dShow at HieQueries.hs:37:1-22, of type: Show (Integer, x, A)
   │     is an evidence variable bound by a let, depending on: [$fShow(,,),
   │                                                            $dShow, $dShow, $dShow]
   │           with scope: LocalScope HieQueries.hs:37:1-22
   │           bound at: HieQueries.hs:37:1-22
   │     Defined at <no location info>
   └
   |
   +- ┌
   |  │ $fShow(,,) at HieQueries.hs:(4,1)-(82,26), of type: forall a b c. (Show a, Show b, Show c) => Show (a, b, c)
   |  │     is an evidence variable bound by an instance
   |  │           with scope: ModuleScope
   |  │           
   |  │     Defined in `GHC.Show'
   |  └
   |
   +- ┌
   |  │ $dShow at HieQueries.hs:37:1-22, of type: Show Integer
   |  │     is an evidence variable bound by a let, depending on: [$fShowInteger]
   |  │           with scope: LocalScope HieQueries.hs:37:1-22
   |  │           bound at: HieQueries.hs:37:1-22
   |  │     Defined at <no location info>
   |  └
   |  |
   |  `- ┌
   |     │ $fShowInteger at HieQueries.hs:(4,1)-(82,26), of type: Show Integer
   |     │     is an evidence variable bound by an instance
   |     │           with scope: ModuleScope
   |     │           
   |     │     Defined in `GHC.Show'
   |     └
   |
   +- ┌
   |  │ $dShow at HieQueries.hs:37:1-22, of type: Show x
   |  │     is an evidence variable bound by a type signature
   |  │           with scope: LocalScope HieQueries.hs:37:1-22
   |  │           bound at: HieQueries.hs:37:1-22
   |  │     Defined at <no location info>
   |  └
   |
   `- ┌
      │ $dShow at HieQueries.hs:37:1-22, of type: Show A
      │     is an evidence variable bound by a let, depending on: [$fShowA]
      │           with scope: LocalScope HieQueries.hs:37:1-22
      │           bound at: HieQueries.hs:37:1-22
      │     Defined at <no location info>
      └
      |
      `- ┌
         │ $fShowA at HieQueries.hs:(4,1)-(82,26), of type: Show A
         │     is an evidence variable bound by an instance
         │           with scope: ModuleScope
         │           
         │     Defined at HieQueries.hs:42:21
         └

